stan/math/rev/fun/multiply_lower_tri_self_transpose.hpp

```diff
@@ -11,43 +11,34 @@
 namespace stan {
 namespace math {
 
-inline matrix_v multiply_lower_tri_self_transpose(const matrix_v& L) {
-  // check_square("multiply_lower_tri_self_transpose",
-  // L, "L", (double*)0);
-  int K = L.rows();
-  int J = L.cols();
-  matrix_v LLt(K, K);
-  if (K == 0) {
-    return LLt;
-  }
-  // if (K == 1) {
-  //   LLt(0, 0) = L(0, 0) * L(0, 0);
-  //   return LLt;
-  // }
-  int Knz;
-  if (K >= J) {
-    Knz = (K - J) * J + (J * (J + 1)) / 2;
-  } else {  // if (K < J)
-    Knz = (K * (K + 1)) / 2;
-  }
-  vari** vs = reinterpret_cast<vari**>(
-      ChainableStack::instance_->memalloc_.alloc(Knz * sizeof(vari*)));
-  int pos = 0;
-  for (int m = 0; m < K; ++m) {
-    for (int n = 0; n < ((J < (m + 1)) ? J : (m + 1)); ++n) {
-      vs[pos++] = L(m, n).vi_;
-    }
-  }
-  for (int m = 0, mpos = 0; m < K; ++m, mpos += (J < m) ? J : m) {
-    LLt.coeffRef(m, m) = var(
-        new internal::dot_self_vari(vs + mpos, (J < (m + 1)) ? J : (m + 1)));
-    for (int n = 0, npos = 0; n < m; ++n, npos += (J < n) ? J : n) {
-      LLt.coeffRef(m, n) = LLt.coeffRef(n, m)
-          = dot_product(L.row(m).head((J < (n + 1)) ? J : (n + 1)),
-                        L.row(n).head((J < (n + 1)) ? J : (n + 1)));
-    }
-  }
-  return LLt;
+template <typename T, require_eigen_vt<is_var, T>* = nullptr>
+inline Eigen::Matrix<var, T::RowsAtCompileTime, T::RowsAtCompileTime>
+multiply_lower_tri_self_transpose(const T& L) {
+  using T_double = promote_scalar_t<double, T>;
+  using T_var = promote_scalar_t<var, T>;
+
+  if (L.rows() == 0)
+    return T_var();
+
+  arena_matrix<T_var> arena_L = L;
+  arena_matrix<T_double> arena_L_val
+      = value_of(arena_L).template triangularView<Eigen::Lower>();
+
+  arena_matrix<T_var> res = arena_L_val.template triangularView<Eigen::Lower>()
+                            * arena_L_val.transpose();
```

Comment:
Another place I think var_value<Matrix> could be used. We would have something like 
```suggestion
  // Should arena_matrix with var<matrix> work?
  arena_matrix<var_matrix_converter_t<T_var>> arena_L = L;
  arena_matrix<T_var> res = arena_L.val().template triangularView<Eigen::Lower>() * arena_L.val().transpose();
```

Reply:
> // Should arena_matrix with var<matrix> work?

That hasn't been done yet. Tadej said he's busy right now. If you want to open the pull, I'm happy to review. I very much want to see these benchmarks with `var_value<Eigen>` types.

Reply:
Cool cool I'll work on that

Reply:
> Tadej said he's busy right now.
I'm not too busy for reviewing or answering questions. I am just not starting another PR like this on my own.

> Should arena_matrix with var<matrix> work?
No. `arena_matrix` is intended for eigen types. `var_value` has allocations on stack anyway, so it can replace `arena_matrix<Eigen something>`. Or you can use `arena_t<>` on either Eigen type or `var_value` and you will get something that can be captured in lambda for `reverse_pass_callback`.

Reply:
> var_value has allocations on stack anyway

Oh right. It's just a pointer.

> I'm not too busy for reviewing or answering questions.

Right right, just letting Steve know that nobody else was working on it.

Reply:
I'll open up a draft PR for this tmrw

> Or you can use arena_t<> on either Eigen type or var_value and you will get something that can be captured in lambda for reverse_pass_callback.

I'll write more when I open the draft, I need to play with it more, but I just wrote a specialiation for `arena_matrix` that just inherits from a `var_value<>` eigen. We might want that style if we want to be able to reuse the same functions for both var<mat> and mat<var> and their combinations

Reply:
> I'll open up a draft PR for this tmrw

Cool. Feel free to take any of the functions here and use in that pull, or softmax is already in: https://github.com/stan-dev/math/blob/develop/stan/math/rev/fun/softmax.hpp